datafusion/physical-expr/src/datetime_expressions.rs

```diff
@@ -396,6 +396,39 @@ fn string_to_timestamp_nanos_shim(s: &str) -> Result<i64> {
     string_to_timestamp_nanos(s).map_err(|e| e.into())
 }
 
+fn to_date_impl(args: &[ColumnarValue], name: &str) -> Result<ColumnarValue> {
+    match args.len() {
+        1 => handle::<Date32Type, _, Date32Type>(
+            args,
+            |s| {
+                string_to_timestamp_nanos_shim(s)
+                    .map(|n| n / (1_000_000 * 24 * 60 * 60 * 1_000))
+                    .and_then(|v| {
+                        v.try_into().map_err(|_| {
+                            internal_datafusion_err!("Unable to cast to Date32 for converting from i64 to i32 failed")
+                        })
+                    })
+            },
+            name,
+        ),
```

Comment:
Maybe we can use `to_timestamp_impl` to avoid code duplication.
```suggestion
        1 => to_timestamp_impl::<TimestampNanosecondType>(args, name),
```

Reply:
We can not do that

for the reason the Date32 need i32 type and time_stamp need i64 type.

to_timestamp_impl did not deal with the converting from i64 to i32